challengeresponse.go

```diff
@@ -66,12 +66,44 @@ func (cfg ChallengeResponseConfig) Run() ([]byte, error) {
 		return nil, fmt.Errorf("evidence generation failed: %w", err)
 	}
 
-	attestationResult, err := cfg.challengeResponse(evidence, mediaType, sessionURI)
+	return cfg.ChallengeResponse(evidence, mediaType, sessionURI)
+}
+
+// NewSession runs the first part of the interaction which deals with session
+// creation, nonce and token format negotiation. On success, the session object
+// is returned together with the URI of the new session endpoint
+func (cfg ChallengeResponseConfig) NewSession() (*RatsChallengeResponseSession, string, error) {
+	if err := cfg.check(false); err != nil {
+		return nil, "", err
+	}
+
+	// Attach the default client if the user hasn't supplied one
+	if cfg.Client == nil {
+		cfg.Client = NewClient()
+	}
+
+	return cfg.newSession()
+}
+
+// ChallengeResponse runs the second portion of the interaction protocol that
+// deals with Evidence submission and retrieval of the associated Attestation
+// Result.  On success, the Attestation result in JSON format is returned.
+func (cfg ChallengeResponseConfig) ChallengeResponse(
+	evidence []byte,
+	mediaType string,
+	uri string,
+) ([]byte, error) {
+	// At this point we must assume we have a Client
+	if cfg.Client == nil {
+		return nil, errors.New("bad configuration: nil client")
+	}
+
+	attestationResult, err := cfg.challengeResponse(evidence, mediaType, uri)
 
 	// if requested, explicitly call DELETE on the session resource
 	if cfg.DeleteSession {
-		if err = cfg.deleteSession(sessionURI); err != nil {
-			log.Printf("DELETE %s failed: %v", sessionURI, err)
+		if err = cfg.deleteSession(uri); err != nil {
+			log.Printf("DELETE %s failed: %v", uri, err)
```

Comment:
If DELETE API fails, should we still return a valid attestationResult? Should the invoker trust the same?

Reply:
> If DELETE API fails, should we still return a valid attestationResult?

yes: DELETE is distinct from the async GET (or the sync POST) that fetches the attestation result.  If DELETE fails (e.g., because the server has already reaped the associated resource), it's perfectly OK from the client point of view.

> Should the invoker trust the same?

I don't understand what "trust the same" refers to.

Reply:
I mean in this use case, there is some error, so the user of the API should it deduce that the Attestation Results are meaningful, as some aspect of the API has failed...

Reply:
ah ok!  yes then, the failed DELETE has no impact on the returned attestation results.